### PR TITLE
cleanup & DRY Queue open(), addIO() and addPage()

### PR DESCRIPTION
**Pull request by colinsurprenant**
This is basically a cleanup of `Queue` `open()` and and DRYing of `addIO()` and `addPage()`. 

The logic is easier to follow in the `open()` method. It is a **long** method but the steps are relatively straight forward plus I added comments for the tricky part.

### CODE REVIEW

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java [hunk 1]
```diff
@@ -250,75 +268,44 @@ public void open() throws IOException {
...
-            this.tailPages.add(page);
-            this.unreadTailPages.add(page);
-            this.unreadCount += page.unreadCount();
-            this.currentByteSize += page.getPageIO().getCapacity();
+    /**
+     * delete files for the given page
+     *
+     * @param checkpoint the tail page {@link Checkpoint}
+     * @param pageIO the tail page {@link PageIO}
+     * @throws IOException
```
Comment (jakelandis) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: nit: IIRC the javadoc tool complains when there is not a description here. that's why you will often see silly comments throws IOException when IO exceptions happen.  

I couldn't find this requirement in the [style doc](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#throwstag), so just an FYI if you see warnings/errors while building javadoc.

**Comment by jakelandis** [comment]
I don't see any new tests, is it safe to assume that existing tests cover this code and didn't require any changes ?

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java [hunk 2]
```diff
@@ -183,12 +183,24 @@ public void open() throws IOException {
             for (int pageNum = headCheckpoint.getFirstUnackedPageNum(); pageNum < headCheckpoint.getPageNum(); pageNum++) {
 
                 // all tail checkpoints in the sequence should exist, if not abort mission with a NoSuchFileException
-                Checkpoint cp = this.checkpointIO.read(this.checkpointIO.tailFileName(pageNum));
+                final Checkpoint cp = this.checkpointIO.read(this.checkpointIO.tailFileName(pageNum));
```
Comment (jakelandis) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: nit: cp --> checkpoint ? (like other variables in the this class)

**Comment by colinsurprenant** [comment]
@jakelandis yes, these changes does not modify the execution logic - it dries up redundant code and makes the open() sequence more understandable. existing tests covers this.

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java [hunk 3]
```diff
@@ -222,17 +234,23 @@ public void open() throws IOException {
...
+                }
 
                 // track the seqNum as we add this new tail page, prevent empty tailPage with a minSeqNum of 0 to reset seqNum
                 if (headCheckpoint.maxSeqNum() > this.seqNum) {
                     this.seqNum = headCheckpoint.maxSeqNum();
                 }
+
+                // create a new empty head page
+                headPageNum = headCheckpoint.getPageNum() + 1;
+                newCheckpointedHeadpage(headPageNum);
```
Comment (jakelandis) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: These two blocks of code swapped position, is that intentional (or doesn't matter)?

FROM
```
               headPageNum = headCheckpoint.getPageNum() + 1;
                newCheckpointedHeadpage(headPageNum);

                // track the seqNum as we add this new tail page, prevent empty tailPage with a minSeqNum of 0 to reset seqNum
                if (headCheckpoint.maxSeqNum() > this.seqNum) {
                    this.seqNum = headCheckpoint.maxSeqNum();
                }
```

TO:
```
                // track the seqNum as we add this new tail page, prevent empty tailPage with a minSeqNum of 0 to reset seqNum
                if (headCheckpoint.maxSeqNum() > this.seqNum) {
                    this.seqNum = headCheckpoint.maxSeqNum();
                }

             headPageNum = headCheckpoint.getPageNum() + 1;
              newCheckpointedHeadpage(headPageNum);
```

Comment (colinsurprenant) [reply logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: yes, it does not matter - I moved the `seqNum` tracking bits for the beheaded head page right after the add/purge page handling to be consistent with the tail pages handling earlier in the code.

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java (hunk 2 above)
Comment (colinsurprenant) [reply logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: meh, will pass on this one ;)

**Comment by colinsurprenant** [comment]
I will followup shortly with a complete javadoc cleanup PR...

**Comment by jakelandis** [comment]
LGTM

**Comment by colinsurprenant** [comment]
merged into master and 6.x. Thanks @jakelandis

**Pull request merged**